torch/_inductor/lowering.py

```diff
@@ -2740,6 +2740,42 @@ def output_indexer(index):
     return self
 
 
+@register_lowering(
+    inductor_prims.masked_scatter_with_index, type_promotion_kind=None, broadcast=False
+)
+def masked_scatter_with_index(self, mask, source_idx, source):
+    self_flat, mask_flat, source_flat = (view(x, (-1,)) for x in (self, mask, source))
+
+    assert self.get_size() == mask.get_size()
+    assert mask.get_dtype() in {torch.bool, torch.uint8}
+
+    self_loader = self_flat.make_loader()
+    mask_loader = mask_flat.make_loader()
+    source_idx_loader = source_idx.make_loader()
+    source_loader = source_flat.make_loader()
+    source_numel = source.get_numel()
+
+    def inner_fn(idx):
+        self_val = self_loader(idx)
+        mask_val = ops.to_dtype(mask_loader(idx), torch.bool)
+
+        def load_source_val():
+            source_idx_val = source_idx_loader(idx)
+            i = ops.indirect_indexing(source_idx_val, source_numel)
+            return source_loader([i])
+
+        source_val = ops.masked(mask_val, load_source_val, 0)
+        return ops.where(mask_val, source_val, self_val)
```

Comment:
In other places in the lowerings we do a `where` after a `masked`, but I think it should be a tad more efficient to do an `ops.add`.

Reply:
`ops.masked` under the hood returns `tl.where(mask, val, 0)`, so by doing another `tl.where` with the same mask value, I expect this to be optimized into a single `tl.where` by the triton compiler.

Reply:
Could you confirm that this is the case? If it is, we should probably change the lowering that does a masked into an add.

Reply:
I'm slightly surprised how far this makes it through the optimization pipeline before getting removed. It goes all the way to the llvm IR:
```mlir
    %74 = llvm.select %44, %73, %8 : i1, f32 loc(#loc27)
    %75 = llvm.select %44, %74, %37 : i1, f32 loc(#loc28)
```

But it is optimized away by the time we get to the ptx.
```ptx
        @%p11 ld.global.b32 { %r9 }, [ %rd23 + 0 ];
        @!%p11 mov.u32 %r9, %r10;
        .loc    1 28 32
        selp.b32        %r11, %r9, %r3, %p6;
        .loc    1 29 25
        shl.b64         %rd26, %rd1, 2;
        add.s64         %rd24, %rd4, %rd26;
        .loc    1 29 36
        @%p13 st.global.b32 [ %rd24 + 0 ], { %r11 };
```

Here you can see only one select op (`selp.b32`) between loading `source_val` (`%r9`) and storing the final result (`%r11`).

> we should probably change the lowering that does a masked into an add.

All I could find was https://github.com/pytorch/pytorch/blob/5531a23b204b4daa2c0bb3c52610e9a0ba79dacf/torch/_inductor/lowering.py#L3199-L3200

which seems to be accumulating a gradient so an add really is what we want. lmk if you had something in mind.

Reply:
yeah, it's true that that `add` is necessary. Thank you for looking into this.

Reply:
I hope the CPU path can also optimise this out (no need to check). It would also be possible to make `masked`  accept a tensor as a second argument as
```python
ops.masked(mask_val, load_source_val, self_val)
```
but yeah, no need.